### PR TITLE
feat(CL): Support Next Tick

### PR DESCRIPTION
**Pull request by alexanderbez**
## Changelog 

* Remove unnecessary `Initialized` field from `TickInfo`
* Index cleanup
  * int64 for ticks
  * sdk.AccAddress for owner with length-prefixing
  * Bytes for prefixes
* Add `NextInitializedTick`

### CODE REVIEW

File: x/concentrated-liquidity/lp.go [hunk 1]
```diff
@@ -8,11 +8,15 @@ import (
...
+func (k Keeper) Mint(ctx sdk.Context, poolId uint64, owner sdk.AccAddress, tokenIn sdk.Int, lowerTick, upperTick int64) (numShares sdk.Int, err error) {
+	// ensure types.MinTick <= lowerTick < types.MaxTick
+	if lowerTick < types.MinTick || lowerTick >= types.MaxTick {
+		return sdk.Int{}, fmt.Errorf("invalid lower tick: %d", lowerTick)
+	}
+
+	// ensure types.MaxTick < upperTick <= types.MinTick
+	if upperTick > types.MaxTick || upperTick <= types.MinTick {
+		return sdk.Int{}, fmt.Errorf("invalid upper tick: %d", upperTick)
 	}
```
Comment (p0mvn) [inline x/concentrated-liquidity/lp.go:20]: Can we add TODO or issue to test this?

File: x/concentrated-liquidity/tick.go [hunk 2]
```diff
@@ -1,36 +1,75 @@
...
 	}
 
-	k.setTickInfo(ctx, poolId, tickIndex, tickInfo)
+	defer iter.Close()
+
+	i := 0
+	for ; iter.Valid() && i < 2; iter.Next() {
+		// Since, we constructed our prefix store with <TickPrefix | poolID>, the
+		// key is the BigEndianToUint64 encoding of a tick index.
+		tick := int64(sdk.BigEndianToUint64(iter.Key()))
```
Comment (p0mvn) [inline x/concentrated-liquidity/tick.go:49]: How are we handling negative ticks?

Comment (p0mvn) [reply x/concentrated-liquidity/tick.go:49]: I think that makes sense now. I missed the int64 cast

File: x/concentrated-liquidity/types/keys.go [hunk 3]
```diff
@@ -1,33 +1,43 @@
...
 
 // KeyPosition uses pool Id, owner, lower tick and upper tick for keys
-func KeyPosition(poolId uint64, address string, lowerTick, upperTick sdk.Int) []byte {
-	return []byte(fmt.Sprintf("%s%d%s%s%s", PositionPrefix, poolId, address, lowerTick.String(), upperTick.String()))
+func KeyPosition(poolId uint64, addr sdk.AccAddress, lowerTick, upperTick int64) []byte {
+	var key []byte
+	key = append(key, PositionPrefix...)
+	key = append(key, address.MustLengthPrefix(addr)...)
+	key = append(key, sdk.Uint64ToBigEndian(uint64(lowerTick))...)
+	key = append(key, sdk.Uint64ToBigEndian(uint64(upperTick))...)
```
Comment (mattverse) [inline x/concentrated-liquidity/types/keys.go:41]: we do have negative ticks though right?

File: x/concentrated-liquidity/tick.go [hunk 4]
```diff
@@ -1,36 +1,75 @@
...
-		tickInfo.Initialized = true
+	k.setTickInfo(ctx, poolId, tickIndex, tickInfo)
+}
+
+// NextInitializedTick returns the next initialized tick index based on the
+// current or provided tick index. If no initialized tick exists, <0, false>
+// will be returned. The lte argument indicates if we need to find the next
+// initialized tick to the left or right of the current tick index, where true
+// indicates searching to the left.
+func (k Keeper) NextInitializedTick(ctx sdk.Context, poolId uint64, tickIndex int64, lte bool) (next int64, initialized bool) {
```
Comment (mattverse) [inline x/concentrated-liquidity/tick.go:27]: nit suggestion: how about reverse?
```suggestion
func (k Keeper) NextInitializedTick(ctx sdk.Context, poolId uint64, tickIndex int64, reverse bool) (next int64, initialized bool) {
```

Comment (mattverse) [reply x/concentrated-liquidity/tick.go:27]: Or reverseSearch

File: x/concentrated-liquidity/tick.go [hunk 5]
```diff
@@ -1,36 +1,75 @@
...
+	store := ctx.KVStore(k.storeKey)
+
+	// Construct a prefix store with a prefix of <TickPrefix | poolID>, allowing
+	// us to retrieve the next initialized tick without having to scan all ticks.
+	prefixBz := types.KeyTickPrefix(poolId)
+	prefixStore := prefix.NewStore(store, prefixBz)
+	startKey := sdk.Uint64ToBigEndian(uint64(tickIndex))
```
Comment (mattverse) [inline x/concentrated-liquidity/tick.go:34]: tickIndex can be negative numbers right?

File: x/concentrated-liquidity/tick.go [hunk 6]
```diff
@@ -1,36 +1,75 @@
...
+		iter = prefixStore.ReverseIterator(nil, startKey)
+	} else {
+		iter = prefixStore.Iterator(startKey, nil)
 	}
 
-	k.setTickInfo(ctx, poolId, tickIndex, tickInfo)
+	defer iter.Close()
+
+	i := 0
+	for ; iter.Valid() && i < 2; iter.Next() {
```
Comment (mattverse) [inline x/concentrated-liquidity/tick.go:46]: why 2? We iterate continuously until we find the next tick right?

File: x/concentrated-liquidity/tick_test.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/testutil"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+)
+
+func Test(t *testing.T) {
```
Comment (mattverse) [inline x/concentrated-liquidity/tick_test.go:11]:
```suggestion
func TestNextInitializedTick(t *testing.T) {
```

We also have test suite set up on main, feel free to use that as well!

File: x/concentrated-liquidity/tick_test.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+			n, initd := k.NextInitializedTick(ctx, 1, 328, false)
+			require.Equal(t, int64(340), n)
+			require.True(t, initd)
+		})
+	})
+
+	t.Run("lte=true", func(t *testing.T) {
+		t.Run("returns tick directly to the left of input tick if not initialized", func(t *testing.T) {
+			n, initd := k.NextInitializedTick(ctx, 1, 79, true)
```
Comment (mattverse) [inline x/concentrated-liquidity/tick_test.go:60]: we should test with values(tick) that is far from the initialized value (e.g100)

**Review by mattverse — CHANGES_REQUESTED** [review]
Other implementations lgtm, requesting change  being unconvinced for the part where we iterate over iterator only twice 
```
	for ; iter.Valid() && i < 2; iter.Next() {
```

File: x/concentrated-liquidity/types/keys.go (hunk 3 above)
Comment (alexanderbez) [reply x/concentrated-liquidity/types/keys.go:41]: Yes! int64 vs uint64 are the same values, they just are represented differently.

E.g.

https://go.dev/play/p/irSTd51l4vz

File: x/concentrated-liquidity/tick.go (hunk 5 above)
Comment (alexanderbez) [reply x/concentrated-liquidity/tick.go:34]: Yup!

Ref: https://go.dev/play/p/irSTd51l4vz

File: x/concentrated-liquidity/tick.go (hunk 6 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/tick.go:48]: Same question

File: x/concentrated-liquidity/tick_test.go (hunk 8 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/tick_test.go:92]: also agreed on this

**Review by czarcas7ic — APPROVED** [review]

File: x/concentrated-liquidity/tick_test.go (hunk 8 above)
Comment (alexanderbez) [reply x/concentrated-liquidity/tick_test.go:91]: Added a test case. LMK if this is what you had in mind

File: x/concentrated-liquidity/tick.go (hunk 6 above)
Comment (alexanderbez) [reply x/concentrated-liquidity/tick.go]: I had it for testing purposes. Removed :)

File: x/concentrated-liquidity/tick.go [hunk 9]
```diff
@@ -1,36 +1,87 @@
...
-		tickInfo.Initialized = true
+	k.setTickInfo(ctx, poolId, tickIndex, tickInfo)
+}
+
+// NextInitializedTick returns the next initialized tick index based on the
+// current or provided tick index. If no initialized tick exists, <0, false>
+// will be returned. The lte argument indicates if we need to find the next
+// initialized tick to the left or right of the current tick index, where true
+// indicates searching to the left.
+func (k Keeper) NextInitializedTick(ctx sdk.Context, poolId uint64, tickIndex int64, lte bool) (next int64, initialized bool) {
```
Comment (alexanderbez) [inline x/concentrated-liquidity/tick.go:30]: @mattverse I kept this as `lte`. After further thought, I think this is more indicative of what's going on and makes it easier to reason about when looking at the method.

File: x/concentrated-liquidity/lp.go (hunk 1 above)
Comment (alexanderbez) [reply x/concentrated-liquidity/lp.go]: Yup!

**Comment by alexanderbez** [comment]
This PR is good to go! Just need to resolve conflicts 🦅

**Review by mattverse — APPROVED** [review]

**Pull request merged**